Distribution of Metrics Publisher #325

mgherghe opened the pull request.

Commit:
Changed the listening port for StatsD; Made minor changes to the Readme file to reflect David's suggestion.

Commit:
Distribution of Metrics Publisher #325:
Created zip distribution with start script for metrics.statsd
Changed logging with slf4j
Created metrics.reader for reading the metrics from Agrona

brennangaunce commented on pom.xml:
```diff
@@ -23,6 +23,7 @@
     </scm>
 
     <modules>
+		<module>reader</module>
```
No tabs.  Please make sure checkstyle is enabled on this project.

mgherghe replied:
I fixed the tabs issue. However, if I enable the checkstyle on this project, the build fails due to the fact that I'm using lambda expressions.

krismcqueen commented on reader/src/main/java/org/kaazing/monitoring/reader/impl/MetricsCollectorAgrona.java:
```diff
@@ -0,0 +1,118 @@
+    public boolean initialize() {
+
+        if (initialized) {
+            LOGGER.debug("Resources already initialized. Exiting method.");
```
typically surrounded with

if (LOGGER.isDebugEnabled()) {
}

to avoid creating a String that will immediately get thrown away when log level does not support DEBUG.

mgherghe replied:
From what I understand from the documentation, slf4j first checks if debug is enabled and only then continues to log the message received as parameter. In this particular case, it is a constant string and there is no additional String created when calling the debug method.

krismcqueen commented on reader/src/main/java/org/kaazing/monitoring/reader/impl/MetricsCollectorAgrona.java:
```diff
@@ -0,0 +1,118 @@
+
+            initialized = true;
+        } catch (Exception e) {
+            LOGGER.debug(e.toString());
```
Surround with LOGGER.isDebugEnabled(), also consider adding the stack trace to the output.

krismcqueen commented on reader/src/main/java/org/kaazing/monitoring/reader/impl/MetricsCollectorAgrona.java:
```diff
@@ -0,0 +1,118 @@
+
+                final int offset = CountersManager.counterOffset(id);
+                final long value = valuesBuffer.getLongVolatile(offset);
+                LOGGER.debug(String.format("%3d: %,10d - %s", id, value, label));
```
see other logging feedback

krismcqueen commented on statsd/src/main/java/org/kaazing/monitoring/client/Configuration.java:
```diff
@@ -0,0 +1,70 @@
+            properties.load(inputStream);
+            inputStream.close();
+        } catch (IOException e) {
+            LOGGER.error(e.getMessage());
```
Consider adding a more descriptive log message and possibly logging the stack trace on DEBUG.

krismcqueen commented on statsd/src/main/resources/assembly/windows-bin.xml:
```diff
@@ -0,0 +1,66 @@
+		</dependencySet>  
+	</dependencySets>  
+  
+</assembly>
```
tabs to spaces in this file

krismcqueen commented on statsd/src/main/resources/assembly/unix-bin.xml:
```diff
@@ -0,0 +1,66 @@
+		</dependencySet>  
+	</dependencySets>  
+  
+</assembly>
```
tabs to spaces in this file

krismcqueen commented on statsd/src/main/java/org/kaazing/monitoring/client/StatsdPublisher.java:
```diff
@@ -115,21 +115,21 @@ private synchronized boolean flush() {
-                log.error(String.format("Could not send entirely stat %s to host %s:%d. Only sent %d bytes out of %d bytes",
-                        sendBuffer.toString(), _address.getHostName(), _address.getPort(), nbSentBytes, sizeOfBuffer));
+                LOGGER.error(String.format("Could not send entirely stat %s to host %s:%d. Only sent %d bytes out of %d bytes",
+                        sendBuffer.toString(), address.getHostName(), address.getPort(), nbSentBytes, sizeOfBuffer));
```
Does sending the incorrect number of bytes represent a complete loss of service, or could this possibly be logged as a warning?

krismcqueen commented on statsd/src/main/java/org/kaazing/monitoring/client/MainApp.java:
```diff
@@ -21,35 +21,69 @@
                 }
 
-                Thread.sleep(1000);
+                Thread.sleep(updateInterval);
```
I think this is probably fine, but if this app is going to be productized then I suggest using something like ScheduledThreadPoolExecutor to run a job every second rather than using Thread.sleep.

krismcqueen commented on statsd/src/main/java/org/kaazing/monitoring/client/MainApp.java:
```diff
@@ -21,35 +21,69 @@
-            final AtomicBoolean running = new AtomicBoolean(true);
+            // Waits until the metrics file is created
+            while (metricsCollector.initialize() == false) {
+                Thread.sleep(updateInterval);
```
What is the asynchronous operation in initialize() that you are waiting for?  When it fails the first time, what will change such that it'll succeed in a subsequent attempt?  I'm not quite understanding this loop, any insights you can provide would be appreciated.

krismcqueen commented:
@mgherghe reviewed.  Ping me when you have had a chance to look at the feedback.

Commit:
Changed tabs to spaces in several files;
Used ScheduledThreadPoolExecutor to run the metrics reading task;
Other minor code changes after code review

krismcqueen commented:
@mgherghe is this ready for review again?

Commit:
Added JavaDocs to several classes
Changed Readme file to reflect the new configuration

mgherghe commented:
@krismcqueen this is now ready for review again. I only have to modify the MetricsViewer class to use the metrics.reader for reading the Agrona counters, but that should be similar to how it's done in the MainApp class.

irina-mitrea-luxoft commented on reader/src/main/java/org/kaazing/monitoring/reader/Configuration.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Class for getting Agrona configuration values
+ */
+public class Configuration {
```
As a future enhancement, we could extract the loadConfiguration and getters in a separate interface implemented by this class.

irina-mitrea-luxoft commented on reader/src/main/java/org/kaazing/monitoring/reader/MetricsCollectorFactory.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Factory class that returns a MetricsCollector instance
+ */
+public class MetricsCollectorFactory {
```
As a future enhancement, we could extract an interface implemented by this class.

irina-mitrea-luxoft commented on reader/src/main/java/org/kaazing/monitoring/reader/Configuration.java:
```diff
@@ -0,0 +1,91 @@
+     * Method that loads the configuration file
+     * @return boolean (false if the file is not loaded properly)
+     */
+    public boolean loadConfigFile() {
```
As a future enhancement, we could consider throwing an exception instead of using a boolean indicating when an exception has occurred. This exception can be added to the interface method prototype and caught in the calling class for additional processing.
